### PR TITLE
Support ribbon service-level rule customization

### PR DESCRIPTION
**Pull request by DerekYRC**
## PR Type

Feature

## Describe what this PR does for and how you did.

Support ribbon service-level rule customization

## Adding the issue link (#xxx) if possible.

https://github.com/Tencent/spring-cloud-tencent/issues/445

## Checklist

- [x] Add copyright holder at the beginning of .class file if it is new.
- [x] Add information of this PR to CHANGELOG.md in root of project.
- [x] All junit tests passing.
- [x] Coverage from `Codecov Report` should not decrease.

## Checklist (Optional)

- [x] Will Pull Request to branch of 2020.0 and 2021.0.
- [x] Add documentation in javadoc in code or comment below the PR if necessary.
- [x] Add your name as @author to the beginning of .class file.

### CODE REVIEW

File: spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import org.springframework.cloud.netflix.ribbon.RibbonAutoConfiguration;
+import org.springframework.cloud.netflix.ribbon.RibbonClient;
+import org.springframework.cloud.netflix.ribbon.RibbonClients;
+import org.springframework.cloud.netflix.ribbon.SpringClientFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+
+/**
+ * @author derekyi
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java:30]: Add `Test for {@link PolarisLoadBalancerCompositeRuleBeanPostProcessor}.`

File: spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ * @date 2022/8/1
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java:31]: Remove the `@data`.

File: spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java [hunk 3]
```diff
@@ -198,4 +206,9 @@ public AbstractLoadBalancerRule getRule() {
 			return new ZoneAvoidanceRule();
 		}
 	}
+
+	@VisibleForTesting
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java:210]: It is already a public method and if `@VisibleForTesting` is necessary？

**Review by lepdou — APPROVED** [review]
LGTM

**Review by misselvexu — DISMISSED** [review]

File: spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java (hunk 3 above)
Comment (misselvexu) [reply spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java:210]: It is not recommended using test function annotations in non-test code .

File: spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java (hunk 1 above)
Comment (DerekYRC) [reply spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java]: Done.

File: spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java (hunk 2 above)
Comment (DerekYRC) [reply spring-cloud-starter-tencent-polaris-router/src/test/java/com/tencent/cloud/polaris/router/beanprocessor/PolarisLoadBalancerCompositeRuleBeanPostProcessorTest.java]: Done.

File: spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java (hunk 3 above)
Comment (DerekYRC) [reply spring-cloud-starter-tencent-polaris-router/src/main/java/com/tencent/cloud/polaris/router/PolarisLoadBalancerCompositeRule.java]: The annotation ```@VisibleForTesting``` on the method ```PolarisLoadBalancerCompositeRule#getDelegateRule``` has been deleted.

**Review by SkyeBeFreeman — APPROVED** [review]

**Comment by SkyeBeFreeman** [comment]
@DerekYRC 刚刚测试发现代码实现有点问题，在`com.tencent.cloud.polaris.router.beanprocessor.PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization`方法中，
```java
List<RouterRequestInterceptor> requestInterceptors = BeanFactoryUtils.getBeans(beanFactory, RouterRequestInterceptor.class);
```
这一行没有正确获取到对应的`RouterRequestInterceptor`，导致路由相关标签无法注入到`PolarisLoadBalancerCompositeRule`中。

**Comment by DerekYRC** [comment]
> @DerekYRC 刚刚测试发现代码实现有点问题，在`com.tencent.cloud.polaris.router.beanprocessor.PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization`方法中，
> 
> ```java
> List<RouterRequestInterceptor> requestInterceptors = BeanFactoryUtils.getBeans(beanFactory, RouterRequestInterceptor.class);
> ```
> 
> 这一行没有正确获取到对应的`RouterRequestInterceptor`，导致路由相关标签无法注入到`PolarisLoadBalancerCompositeRule`中。

我研究下

**Comment by DerekYRC** [comment]
> @DerekYRC 刚刚测试发现代码实现有点问题，在`com.tencent.cloud.polaris.router.beanprocessor.PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization`方法中，
> 
> ```java
> List<RouterRequestInterceptor> requestInterceptors = BeanFactoryUtils.getBeans(beanFactory, RouterRequestInterceptor.class);
> ```
> 
> 这一行没有正确获取到对应的`RouterRequestInterceptor`，导致路由相关标签无法注入到`PolarisLoadBalancerCompositeRule`中。

@SkyeBeFreeman 

问题出现在NamedContextFactory#createContext中创建的子容器没有继承父容器的bean，很奇怪子容器获取不到父容器的bean，有时间深究下，可能是框架的bug。

一种解决方法是PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization方法中除IClientConfig外的所有bean都从父容器中获取。

今天修复下。

**Comment by SkyeBeFreeman** [comment]
> > @DerekYRC 刚刚测试发现代码实现有点问题，在`com.tencent.cloud.polaris.router.beanprocessor.PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization`方法中，
> > ```java
> > List<RouterRequestInterceptor> requestInterceptors = BeanFactoryUtils.getBeans(beanFactory, RouterRequestInterceptor.class);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这一行没有正确获取到对应的`RouterRequestInterceptor`，导致路由相关标签无法注入到`PolarisLoadBalancerCompositeRule`中。
> 
> @SkyeBeFreeman
> 
> 问题出现在NamedContextFactory#createContext中创建的子容器没有继承父容器的bean，很奇怪子容器获取不到父容器的bean，有时间深究下，可能是框架的bug。
> 
> 一种解决方法是PolarisLoadBalancerCompositeRuleBeanPostProcessor#postProcessBeforeInitialization方法中除IClientConfig外的所有bean都从父容器中获取。
> 
> 今天修复下。

是不是可以做个判断，如果子容器管理没有找到，就去父容器管理找找？

**Comment by DerekYRC** [comment]
@SkyeBeFreeman 找到子容器获取不到父容器的bean的原因了，今晚再详细说明和修复下

**Comment by DerekYRC** [comment]
@SkyeBeFreeman PR：https://github.com/Tencent/spring-cloud-tencent/pull/515

**Pull request merged**